src/mpid/ch4/netmod/ofi/ofi_win.c

```diff
@@ -547,13 +552,15 @@ static void dwin_close_mr(void *obj)
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_DWIN_CLOSE_MR);
 
     /* Close local MR */
-    struct fid_mr *mr = (struct fid_mr *) obj;
-    if (mr) {
+    dwin_origin_mr_t *origin_mr = (dwin_origin_mr_t *) obj;
+    if (origin_mr) {
         int ret;
-        MPIDI_OFI_CALL_RETURN(fi_close(&mr->fid), ret);
+        MPIDI_OFI_mr_key_free(origin_mr->requested_key);
+        MPIDI_OFI_CALL_RETURN(fi_close(&origin_mr->mr->fid), ret);
         MPIR_Assert(ret >= 0);
     }
```

Comment:
We need free `requested_key` only when `MPIDI_OFI_ENABLE_MR_PROV_KEY` is off. Besides, we can get the key via mr (i.e., `fi_mr_key(mr)`), thus we don't need define `dwin_origin_mr_t`.

You can refer to the free logic in `MPIDI_OFI_am_rdma_read_ack_handler`.
```
    if (!MPIDI_OFI_ENABLE_MR_PROV_KEY) {
        uint64_t mr_key = fi_mr_key(MPIDI_OFI_AMREQUEST_HDR(sreq, lmt_mr));
        MPIDI_OFI_mr_key_free(mr_key);
    }
```

Reply:
Yeah, I agree we only need to free `requested_key` when `MPIDI_OFI_ENABLE_MR_PROV_KEY` is off.
But I am worried that `fi_mr_key(mr)` does not necessarily mean the `requested_key` generated by MPICH, because you cannot guarantee `fi_mr_key(mr)` == `requested_key`, it depends on how the ofi implements it.

Reply:
I checked the [OFI document](https://ofiwg.github.io/libfabric/v1.11.2/man/fi_mr.3.html), it does not say that the provider may change user-specific `requested_key`. Thus, as long as `fi_mr_reg` succeeds, we can assume `requested_key == fi_mr_key(mr)`. We always assume it is true in MPICH. You can search code that uses `MPIDI_OFI_mr_key_free`.

Reply:
ok, got it. I can fix it